Replace os.path.* with pathlib

eumiro opened the pull request.

petri commented:
By the way, please can you include changelog entries into HISTORY.rst in PRs so that you get proper credit for your work?

jugmac00 commented on tests/test_decoding.py:
```diff
@@ -64,56 +65,52 @@ def objcodes(tnef):
+    zip_data = to_zip((datadir / "one-file.tnef").read_bytes())
+    with tempfile.TemporaryFile() as out:
+        out.write(zip_data)
 
```
I just noticed there is no assert statement. While it missed before, maybe you can add an assert statement? ie peek into the zip file and assert the contained file is present?

I used something like this in my other project's test suite:
```
    zip_stream = io.BytesIO(byte..)
    zip_file = zipfile.ZipFile(zip_stream)
    assert zip_file.namelist() == ['carm.lic']
```

eumiro replied:
Thank you for the hint.

jugmac00 commented on tnefparse/tnef.py:
```diff
@@ -402,8 +401,7 @@ def to_zip(data, default_name='no-name', deflate=True):
             # uniqify this file name by adding -<num> before the extension
-            root, ext = os.path.splitext(filename)
-            tozip[filename].append((attachment.data, str("%s-%d%s" % (root, L + 1, ext))))
+            tozip[filename].append((attachment.data, filename.with_stem(f"{filename.stem}-{L + 1}")))
```
As this line is not covered by a test, this would be a great chance to do so :-)

eumiro replied:
Added an extra test.

jrideout commented:
@eumiro Would you mind rebasing and fixing this PR now that we've made some conflicting changes (including merging your other PRs) ?

eumiro commented:
Yes, @jrideout. The problem was Python 3.9's new method `with_stem` which I could replace with `with_name`.

Now I reviewed and reworked the whole PR and added `tmp_path` fixtures that provide a `pathlib.Path` in a central temporary directory structure.

jrideout commented:
@eumiro Can you please rebase once more?

eumiro commented:
@jrideout done
